File: LLD/Design-Patterns/Singleton/code/Singleton.cpp
// Singleton Pattern — thread-safe lazy initialization (C++11 "magic static").
// Compile: g++ -std=c++11 -pthread Singleton.cpp -o singleton
// Run:     ./singleton

#include <iostream>
#include <mutex>

class Singleton {
public:
    // Deleted copy operations so no one can copy the single instance
    Singleton(const Singleton&) = delete;
    Singleton& operator=(const Singleton&) = delete;

    // C++11 guarantees thread-safe initialization of function-local statics
    static Singleton& getInstance() {
        static Singleton instance;
        return instance;
    }

    void logRequest() {
        requestCount++;
        std::cout << "Handled request #" << requestCount << std::endl;
    }

private:
    int requestCount = 0;
    Singleton() {
        std::cout << "Singleton instance created." << std::endl;
    }
};

int main() {
    Singleton& a = Singleton::getInstance();
    Singleton& b = Singleton::getInstance();

    a.logRequest();
    b.logRequest();

    std::cout << "&a == &b: " << (&a == &b ? "true" : "false") << std::endl;
    return 0;
}

File: LLD/Design-Patterns/Singleton/code/Singleton.java
/**
 * Singleton Pattern — thread-safe lazy initialization (double-checked locking).
 * Compile:  javac Singleton.java
 * Run:      java Singleton
 */
public class Singleton {

    // volatile ensures visibility of the fully-constructed instance across threads
    private static volatile Singleton instance;

    private int requestCount = 0;

    // Private constructor prevents `new Singleton()` from outside this class
    private Singleton() {
        System.out.println("Singleton instance created.");
    }

    public static Singleton getInstance() {
        if (instance == null) {                 // first check (no locking, fast path)
            synchronized (Singleton.class) {
                if (instance == null) {          // second check (inside the lock)
                    instance = new Singleton();
                }
            }
        }
        return instance;
    }

    public void logRequest() {
        requestCount++;
        System.out.println("Handled request #" + requestCount);
    }

    public static void main(String[] args) {
        Singleton a = Singleton.getInstance();
        Singleton b = Singleton.getInstance();

        a.logRequest();
        b.logRequest();

        System.out.println("a == b: " + (a == b)); // true -> same instance
    }
}
